packages/data-service/src/data-service.ts

```diff
@@ -1996,6 +2033,44 @@ export class DataServiceImpl extends EventEmitter implements DataService {
     return !!this._metadataClient;
   }
 
+  isOperationCancelledError(error: Error): boolean {
+    return error.name === OPERATION_CANCELLED_ERROR;
+  }
+
+  private async cancellableOperation<T>(
+    start: () => Promise<T>,
+    stop: () => Promise<void>,
+    abortSignal?: AbortSignal
+  ): Promise<T> {
+    if (!abortSignal) {
+      return await start();
+    }
+
+    if (abortSignal.aborted) {
+      return Promise.reject(createCancelPromiseError());
+    }
+
+    const session = this.startSession('CRUD');
+    const abort = () => {
+      Promise.all([stop(), this.killSessions(session)]).catch((err) => {
+        log.warn(
+          mongoLogId(1_001_000_140),
+          'CancelOp',
+          'Attempting to kill the session failed',
+          { error: err.message }
+        );
+      });
+    };
+    abortSignal.addEventListener('abort', abort, { once: true });
+    let result: T;
+    try {
+      result = await raceWithAbort(start(), abortSignal);
```

Comment:
This is more a personal opinion .. I don't love all that much that we are adding this external `raceWithAbort` and "cancellable promise", is super neat to put that in a different file, but unless I'm missing something is not necessary and creates a new "Concept" that I'd have a hard time to wrap my head around. 

Unless I misunderstood what that does, the whole "raceWithAbort can be replaced here by this code directly:

```
let abortListener;

const aborted = new Promise(,reject=> {
  abortListener = reject;
  abortController.addEventListener('abort', () => reject(new OperationCancelledError()), {once: true});
});

try {
  await Promise.race([aborted, promise]);
} finally {
  abortController.removeEventListener('abort', abortListener);
}
```

Reply:
@mcasimir thanks for pairing up on this one and helping me clean it!